src/core/scheduler.cc

```diff
@@ -65,34 +81,44 @@ Scheduler::Scheduler() {
   }
   visualization_ = VisualizationAdaptor::Create(param->visualization_engine_);
   root_visualization_ = new RootAdaptor();
-  bound_space_ = new BoundSpace();
-  displacement_ = new DisplacementOp();
-  diffusion_ = new DiffusionOp();
-
-  auto* first_op = new FirstOp();
-  auto* biology_module_op = new BiologyModuleOp();
-  auto* discretization_op = new DiscretizationOp();
-  auto* last_op = new LastOp();
-  operations_ = {first_op,      bound_space_,      biology_module_op,
-                 displacement_, discretization_op, last_op};
-  protected_operations_ = {first_op->name_, biology_module_op->name_,
-                           discretization_op->name_, last_op->name_};
+
+  default_ops_ = {"first op",     "bound space",    "biology module",
+                  "displacement", "discretization", "diffusion",
+                  "last op"};
+
+  for (auto& def_op : default_ops_) {
+    ScheduleOp(NewOperation(def_op));
+  }
+
+  protected_ops_ = {"first op", "biology module", "discretization", "last op"};
 }
 
 Scheduler::~Scheduler() {
-  delete backup_;
-  delete root_visualization_;
-  delete diffusion_;
-  for (auto* op : operations_) {
+  for (auto* op : unscheduled_ops_) {
+    delete op;
+  }
+  for (auto* op : scheduled_row_wise_ops_) {
+    delete op;
+  }
+  for (auto* op : scheduled_column_wise_ops_) {
     delete op;
   }
+  // Normally this list of ops is empty, but it can happen that an uninitialized
+  // Scheduler gets destructed, leaving all the ops in this vector
+  // In such cases we loop over this vector
+  for (auto* op : ops_to_add_) {
+    delete op;
+  }
+  delete backup_;
+  delete root_visualization_;
   if (visualization_) {
     delete visualization_;
   }
   auto* param = Simulation::GetActive()->GetParam();
   if (param->statistics_) {
     std::cout << gStatistics << std::endl;
   }
+  // We do not delete operations here; we leave that to OperationRegistry
```

Comment:
Since we always clone operations, doesn't this cause a memory leak?

Reply:
Yes, this is a stale comment from a previous implementation where we didn't clone operations. See a few lines above, where we actually delete all the scheduled operations.